trieste/utils/pareto.py

```diff
@@ -0,0 +1,309 @@
+        self.update()
+
+    @staticmethod
+    def _is_test_required(smaller: tf.Tensor) -> tf.Tensor:
+        """
+        Tests if a point augments or dominates the Pareto set.
+        :param smaller: a boolean tf.Tensor storing test point < Pareto front
+        :return: True if the test point dominates or augments the Pareto front (boolean)
+        """
+        # if and only if the test point is at least in one dimension smaller
+        # for every point in the Pareto set
+        idx_dom_augm = tf.reduce_any(smaller, axis=1)
+        is_dom_augm = tf.reduce_all(idx_dom_augm)
+
+        return is_dom_augm
+
+    def _update_front(self) -> tf.Tensor:
+        """
+        Calculate the non-dominated set of points based on the latest data.
+        The stored Pareto set is sorted on the first objective in ascending order.
+        :return: boolean, whether the Pareto set has actually changed since the last iteration
+        """
+        current = self.front
+        pf, _ = non_dominated_sort(self.data)
+        self.front = tf.gather_nd(pf, tf.expand_dims(tf.argsort(pf[:, 0]), axis=1))
+        return not np.array_equal(current.numpy(), self.front.numpy())
+
+    def update(self, datasets: Mapping[str, Dataset] = None, generic_strategy=False):
+        """
+        Update with new output data.
+        Computes the Pareto set and if it has changed recalculates
+        the cell bounds covering the non-dominated region.
+        For the latter, a direct algorithm is used for two objectives, otherwise a
+        generic divide and conquer strategy is employed.
+        :param datasets: Mapping of Dataset and its coresponding string key(s)
+        :param generic_strategy: Force the generic divide and conquer strategy
+            regardless of the number of objectives (default False)
+        """
+
+        if datasets is not None:
+            data = to_Dataset(datasets)
+        self.data = data if datasets is not None else self.data
+
+        # Find (new) set of non-dominated points
+        changed = self._update_front()
+
+        # Recompute cell bounds if required
+        # Note: if the Pareto set is based on model predictions
+        # it will almost always change in between optimizations
+        if changed:
+            # Clear data container
+            self.bounds.clear()
+            if generic_strategy:
+                self.divide_conquer_nd()
+            else:
+                self.bounds_2d() if self.data.observations.shape[
+                    1
+                ] == 2 else self.divide_conquer_nd()
+
+    def divide_conquer_nd(self):
+        """
+        Divide and conquer strategy to compute the cells covering the non-dominated region.
+        Generic version: works for an arbitrary number of objectives.
+        """
+        outdim = self.data.observations.shape[1]
+
+        # The divide and conquer algorithm operates on a pseudo Pareto set
+        # that is a mapping of the real Pareto set to discrete values
```

Comment:
comments degrade over time, and can very quickly become counter-productive. Can we leave them out?